### PR TITLE
Add deletion

### PR DESCRIPTION
**Pull request by alonre24**
This PR introduces the implementation of our new deletion algorithm for hnsw index.

It includes:
- Refactor of hnswalg.h - removing irrelevant functions, adding documentation, making code more readable.
- Add a build argument "ENABLE_PARALLELISATION", that enables synchronisation in the code for parallel indexing / searching (as hnsw supported before). Note that adding a new element to the index along with searching is not supported, and that removing (or updating data) of an element cannot be done in parallel with any other operation.
- Change example test to receive argument via command line, and to perform arbitrary deletion after every 10 insert operation. Also, print time and memory stats whenever index size is a power of 10.
- Add python bindings for the bruteforce algorithm as well, to allow measuring recall when using python API.

### CODE REVIEW

File: hnswlib/hnswalg.h [hunk 1]
```diff
@@ -8,42 +8,110 @@
...
+        // Index data structures
+        tableint enterpoint_node_;
+        char *data_level0_memory_;
+        char **linkLists_;
+        std::vector<int> element_levels_;
+        std::set<tableint> available_ids;
+        std::unordered_map<labeltype, tableint> label_lookup_;
+        VisitedListPool *visited_list_pool_;
+
+        // used for synchronization only when parallel indexing / searching is enabled.
```
Comment (DvirDukhan) [inline hnswlib/hnswalg.h:71]: `#ifdef PRALLEL_HNSW`

File: hnswlib/hnswalg.h [hunk 2]
```diff
@@ -8,42 +8,110 @@
...
+
+        HierarchicalNSW(SpaceInterface<dist_t> *s, const std::string &location, size_t max_elements=0) {
             loadIndex(location, s, max_elements);
         }
 
-        HierarchicalNSW(SpaceInterface<dist_t> *s, size_t max_elements, size_t M = 16, size_t ef_construction = 200, size_t random_seed = 100) :
-                link_list_locks_(max_elements), link_list_update_locks_(max_update_element_locks), element_levels_(max_elements) {
-            max_elements_ = max_elements;
+        HierarchicalNSW(SpaceInterface<dist_t> *s, size_t max_elements, size_t M = 16, size_t ef_construction = 200, size_t ef = 10, size_t random_seed = 100) :
+                link_list_locks_(max_elements), element_levels_(max_elements) {
```
Comment (DvirDukhan) [inline hnswlib/hnswalg.h]: add `link_list_locks` to `IFDEF PARALLEL_HNSW`

File: hnswlib/hnswalg.h [hunk 3]
```diff
@@ -52,87 +120,40 @@ namespace hnswlib {
...
-            //initializations for special treatment of the first node
-            enterpoint_node_ = -1;
-            maxlevel_ = -1;
 
             linkLists_ = (char **) malloc(sizeof(void *) * max_elements_);
             if (linkLists_ == nullptr)
                 throw std::runtime_error("Not enough memory: HierarchicalNSW failed to allocate linklists");
-            size_links_per_element_ = maxM_ * sizeof(tableint) + sizeof(linklistsizeint);
-            mult_ = 1 / log(1.0 * M_);
-            revSize_ = 1.0 / mult_;
```
Comment (DvirDukhan) [inline hnswlib/hnswalg.h:70]: deleted since not in use

File: hnswlib/hnswalg.h [hunk 4]
```diff
@@ -8,42 +8,110 @@
...
+
+            mult_ = 1 / log(1.0 * M_);
             level_generator_.seed(random_seed);
-            update_probability_generator_.seed(random_seed + 1);
 
-            size_links_level0_ = maxM0_ * sizeof(tableint) + sizeof(linklistsizeint);
+            // data_level0_memory will look like this:
+            // -----4------ | -----4*M0----------- | ----8------------------| ------32------- | ----8---- |
+            // <links_len>  | <link_1> <link_2>... | <incoming_links_set> |   <data>        |  <label>
+            size_links_level0_ = maxM0_ * sizeof(tableint) + sizeof(linklistsizeint)
```
Comment (DvirDukhan) [inline hnswlib/hnswalg.h]: reorder according to the documenation

File: hnswlib/hnswalg.h [hunk 5]
```diff
@@ -52,87 +120,40 @@ namespace hnswlib {
...
+            // -----4------ | -----4*M-------------- | ----8------------------|
+            // <links_len>  | <link_1> <link_2> ...  | <incoming_links_set>
+            size_links_per_element_ = maxM_ * sizeof(tableint) + sizeof(linklistsizeint)
+              + sizeof(void *);
+            incoming_links_offset = maxM_ * sizeof(tableint) + sizeof(linklistsizeint);
+        }
 
         ~HierarchicalNSW() {
 
-            free(data_level0_memory_);
```
Comment (DvirDukhan) [inline hnswlib/hnswalg.h:82]: restore this delete

File: hnswlib/hnswalg.h [hunk 6]
```diff
@@ -158,26 +179,57 @@ namespace hnswlib {
...
+                       &set_ptr, sizeof(void*));
+            }
+        }
+
+
+        linklistsizeint *get_linklist0(tableint internal_id) const {
+            return (linklistsizeint *) (data_level0_memory_ + internal_id * size_data_per_element_ + offsetLevel0_);
+        };
+
+        linklistsizeint *get_linklist0(tableint internal_id, char *data_level0_memory_) const {
```
Comment (DvirDukhan) [inline hnswlib/hnswalg.h]: remove

File: hnswlib/hnswalg.h [hunk 7]
```diff
@@ -158,26 +179,57 @@ namespace hnswlib {
...
-        searchBaseLayer(tableint ep_id, const void *data_point, int layer) {
+        linklistsizeint *get_linklist(tableint internal_id, int level) const {
+            return (linklistsizeint *) (linkLists_[internal_id] + (level - 1) * size_links_per_element_);
+        };
+
+        linklistsizeint *get_linklist_at_level(tableint internal_id, int level) const {
+            return level == 0 ? get_linklist0(internal_id) : get_linklist(internal_id, level);
+        };
+
+        CandidatesQueue<dist_t> searchBaseLayer(tableint ep_id, const void *data_point, int layer) {
```
Comment (DvirDukhan) [inline hnswlib/hnswalg.h]: merge with seatchBaseLayer

File: hnswlib/hnswalg.h [hunk 8]
```diff
@@ -584,477 +546,316 @@ namespace hnswlib {
...
         std::vector<data_t> getDataByLabel(labeltype label)
         {
             tableint label_c;
-            auto search = label_lookup_.find(label);
-            if (search == label_lookup_.end() || isMarkedDeleted(search->second)) {
+            if (label_lookup_.find(label) == label_lookup_.end()) {
                 throw std::runtime_error("Label not found");
             }
-            label_c = search->second;
+            label_c = label_lookup_[label];
```
Comment (DvirDukhan) [inline hnswlib/hnswalg.h]: this is another retrival of the data that you got in line 561

File: hnswlib/hnswalg.h [hunk 9]
```diff
@@ -584,477 +546,316 @@ namespace hnswlib {
...
-                    sNeigh.insert(elOneHop);
+            // check the diff in the link list, save the neighbours that were chosen to be removed, and update the new neighbours
+            tableint removed_links[neighbour_neighbours_count];
+            size_t removed_idx=0;
+            size_t link_idx=0;
 
-                    std::vector<tableint> listTwoHop = getConnectionsWithLock(elOneHop, layer);
-                    for (auto&& elTwoHop : listTwoHop) {
-                        sCand.insert(elTwoHop);
+            while (orig_candidates.size() > 0) {
```
Comment (DvirDukhan) [inline hnswlib/hnswalg.h]: try to put duplicated code in a function

**Review by DvirDukhan — CHANGES_REQUESTED** [review]
also add `update` function please

File: hnswlib/hnswalg.h (hunk 3 above)
Comment (alonre24) [reply hnswlib/hnswalg.h:70]: **only `revSize`**

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**